source/semaphore.c

```diff
@@ -74,7 +74,10 @@ void aws_semaphore_release(struct aws_semaphore *semaphore, size_t num_resources
     if (semaphore->count > semaphore->max_count) {
         semaphore->count = semaphore->max_count;
     }
-    aws_condition_variable_notify_one(&semaphore->sync_point);
+
+    for (size_t i = 0; i < num_resources; ++i) {
+        aws_condition_variable_notify_one(&semaphore->sync_point);
```

Comment:
I wonder if just using a mutex instead of a semaphore to wake up readers would be better? It would avoid the need to loop over notify_one.

Reply:
This would be a different protocol at this point. A sketch:

```
struct rwlock {
  // If negative, a write lock is pending; readers = (remaining readers) - max_readers
  // If non-negative, no write lock is pending; readers = number of read locks
  atomic readers = 0;
  mutex mut_writer;
  // True if a writer is active; this blocks other writers from entering while 
  // a write lock is pending
  bool writer_active;
  // signalled when a write lock is released to allow readers to proceed
  cvar release_readers;
  // signalled when the last reader leaves the lock while a write is pending
  cvar reader_to_writer;
  // signalled once when a write lock is released to allow one pending writer to proceed
  cvar writer_to_writer;
};

read_lock:
  loop {
    compare_exchange: if readers >= 0 { readers++; return; }
    lock (mut_writer) {
      while (readers < 0) { await release_readers; }
    }
 }

read_unlock:
    new_val = readers--;
    if (new_val == -max_readers) {
      // this lock is critical to avoid a race where the writer is just before the sleep when we update readers
      lock (mut_writer) { cvar_awake(reader_to_writer); }
    }

write_lock:
  lock(mut_writer);
  while (writer_active) { await(writer_to_writer); }
  writer_active = true;
  atomic: readers -= max_readers;
  while (readers > -max_readers) {
    await(reader_to_writer);
  }

write_unlock:
  writer_active = false;
  readers += max_readers;
  cvar_awake_all(release_readers);
  cvar_awake(writer_to_writer);
  unlock(mut_writer);  

```

Reply:
Okay so maybe I just don't understand the pseudo-code syntax, but here are my thoughts so far:
1. If there's an open reader, and a new reader and a new writer come along, mut_writer is locked by write_lock and read_lock, as far as I can tell it's up to the scheduler whether the next holder of the lock will be a writer or the readers.
2. If readers are open and someone calls write_lock, acquiring mut_writer, how would read_unlock acquire mut_writer to notify the cvar?
3. I don't at all understand the read_lock function, is that supposed to be like:
```c
while (true) {
    num_readers = aws_atomic_load_int(&lock->readers);
    if (num_readers >= 0) {
        if (aws_atomic_compare_exchange_int(&lock->readers, (size_t *)&num_readers, num_readers + 1)) {
            return success;
        }
    } else {
        lock(mut_writer);
        cvar_wait_pred(release_readers, [&]() { aws_atomic_load_int(&lock->readers) >= 0; });
    }
}
```
Or am I missing something?

I'm sure there's a lot more going on here, tomorrow I'll take a crack at implementing this and see how it goes.